lib/helpers.js

```diff
@@ -127,11 +130,58 @@ const parseResults = (editor, results, filePath, showIgnored) => {
   return toReturn;
 };
 
+function isDirectory(dirPath) {
+  try {
+    return fs.statSync(dirPath).isDirectory();
+  } catch (e) {
+    return false;
+  }
+}
+
+export function findStylelintDirectory(modulesDir) {
+  const stylelintDir = path.join(modulesDir || '', 'stylelint');
+  if (isDirectory(stylelintDir)) {
+    return stylelintDir;
+  }
+  return stylelintLocalPath;
+}
+
+export function getStylelintFromDirectory(modulesDir) {
+  const stylelintDir = findStylelintDirectory(modulesDir);
+  try {
+    // eslint-disable-next-line import/no-dynamic-require
+    return require(stylelintDir);
+  } catch (e) {
+    if (e.code === 'MODULE_NOT_FOUND') {
+      throw new Error('Stylelint not found, try restarting Atom to clear caches.');
+    }
+    // eslint-disable-next-line import/no-dynamic-require
+    return require(stylelintLocalPath);
+  }
+}
+
+export function refreshModulesPath(modulesDir) {
+  if (lastModulesPath !== modulesDir) {
+    lastModulesPath = modulesDir;
+    process.env.NODE_PATH = modulesDir || '';
+    // eslint-disable-next-line no-underscore-dangle
+    require('module').Module._initPaths();
+  }
+}
+
+export function getStylelintInstance(filePath) {
+  const fileDir = path.dirname(filePath);
+  const modulesDir = path.dirname(findCached(fileDir, 'node_modules/stylelint') || '');
```

Comment:
This should probably also use `require.resolve`.

Reply:
I think ```require.resolve``` isn't appropriate here as ```modulesDir``` needs to be the modules directory and not the path of a resolved stylelint module. Changing this to ```const modulesDir = require.resolve('stylelint', { paths: [fileDir] });``` resulted in this test failure:
```
Helpers
  getStylelintInstance && getStylelintFromDirectory
    it tries to find a local stylelint
      Expected Function to be 'located'.
        Error: Expected Function to be 'located'.
        at linter-stylelint/spec/helpers-spec.js:30:25
        at jasmine.Spec.middleware (linter-stylelint/node_modules/jasmine-fix/index.js:49:21)

```

Reply:
I meant a `path.dirname(require.resolve(...))`, guess I should have been a bit more specific sorry 😉.

Reply:
The idea is to reuse the `require` machinery as much as possible so it follows what would "normally" happen.

Reply:
Makes sense, will do!

Reply:
After trying to get `require.resolve` to work I don't think we want to use it to get `modulesDir`. We're trying to find the `node_modules` directory that contains the local stylelint. There's a couple problems with using `require.resolve` for this:

1. `require.resolve` returns the full path to the module file, in the case of stylelint, that's `'node_modules/stylelint/lib/index.js'`. So to get just the path for `node_modules` we need to remove `'/stylelint/lib/index.js'` from everything require.resolve returns.
2. In order to get `require.resolve` to find a local stylelint and not the plugin's stylelint, we need to provide a path to a `node_modules` directory that it can search. However, that path is exactly what we're trying to find, and if we had it already then we don't need to use `require.resolve`.

I think using `findCached` is a pretty good solution, given those problems with `require.resolve`, and it's the same method [used in `linter-eslint`](https://github.com/AtomLinter/linter-eslint/blob/master/src/worker-helpers.js#L112).

Reply:
Any thoughts on this @Arcanemagus ?

Reply:
Hmmm, I thought I'd responded here already, but apparently I'd only _thought_ about it:

1. It's admittedly a bit awkward, but I still think this is what I would go with:
    ```js
    const stylelintPath = require.resolve('stylelint', { paths: [fileDir] });
    // Comment describing what is going on
    const modulesDir = path.dirname(path.dirname(path.dirname(resPath)));
    ```
    Again, the primary motivation is to use Node.js's own mechanics, the ones that will be doing the actual requiring _anyway_, in order to find the module. Maybe with some refactoring/passing that found path around so it doesn't need to find it again in `getStylelintFromDirectory`?
    * Although the `findCached` way of doing it _looks_ a bit cleaner, it's a major hack and can find things that `require` normally wouldn't. `linter-eslint` is only using it because that's what we came up with at the time 😛.
2. We aren't actually passing in a `node_modules` path, we're passing in the path of the directory of the file we are linting and letting `require.resolve` find the right `node_modules` for us.

Reply:
So I haven't been able to get the test to pass using the suggested code. The require mechanics don't seem to find the local stylelint instance when using `fileDir` as the path. I'm currently experimenting with what it would take to get things to work while still using `require.resolve`, including digging into the `Module` source code a bit to understand how require is actually working. I'll update here when I find out more.

Reply:
Oh really? Odd. I pulled that code from a breakpoint within the specs but never actually _ran_ the specs with it. I'll try that out myself.

Reply:
@Arcanemagus I was finally able to get this working by using [browserify/resolve](https://github.com/browserify/resolve) in 11b1d73. This is the same method [used by grunt-cli](https://github.com/gruntjs/grunt-cli/blob/master/bin/grunt).